### ISSUE TITLE
Native Resolution for PSP games on PSVITA. [$2,500]

### ISSUE THREAD

**Issue by Epsilon1033** [vita-nuova/bounties#5]
As the title says, Native resolution for PSP games. (960x544). PSP games are limited to half the vita's native resolution, so they don't exactly look great. The feasibility has already been discussed on reddit so it is certainly possible; its just alot of work.

From reddit (/u/MrColdBird):

"Technically a GE chain hook and a vita-sided port of the PPSSPP renderer would already be enough to get "perfect" PSP emulation with upgraded visuals."

"Let the MIPS processor do the whole PSP work... And tunnel out the GE render chain into VitaOS via kermit, then render it there."

"The GE chain and all registers for buffer transfer are well documented in several PSP developer wikis as well as the PPSSPP sourcecode."

Same goes for kermit on Vita and it's message structures. All it takes is someone that can be assed to put the pieces together."

**Comment by robots** [comment]
For less technical audience I'll add description of Vita's hardware:
- Vita's CPU is 4 cores of ARM Cortex-A9 + 4 cores of PowerVR GPU + 8 cores for video/audio codec processing
- Sony decided to include PSP's CPU (mips) + GE gpu inside Vita's hardware

All PSP games run natively on PSP hardware with ME(media engine) hardware being emulated by Vita's cpu.

Goal of this project is to emulate GE and redirect the drawing to Vita's GPU.

**Comment by robots** [comment]
I have started working on this. But do not expect very fast development. This issue is very complicated one in terms of "time". As said in the description everything is documented - but its not only about documentation, its also about understanding and putting pieces together.

So the roadmap is something like this:
1. patch PSP's sceGe module to redirect Texture memory/DrawList memory to shared Vita memory instead of GE's private memory
2. port PPSSPP's GE implementation to Vita - this step will not include scaling for sake of simplicity
3. add scaling to GE implementation once 2. works correctly.

Point 2. is the hardest one here, as some games are timing dependant and it will be tricky to get it right. PPSSPP contains lot of fixes for this, but Vita+ePSP is a bit different in this matter - 2 separate CPUs.

**Comment by and-war** [comment]
Would this then upscale PSX/PS1 games played on PSVita too? would be great to see them rendered in Higher deff (960x544).

**Comment by Rinnegatamante** [comment]
Technically this could be possible by running PSX games through POPS via PSPEMU i think.

**Comment by 1mpl0de** [comment]
hey @robots how is it coming along? any progress?

**Comment by robots** [comment]
I have had busy time with work + vacation + busy month ahead of me.

As soon as i have something worth seeing I will create repo on github to track the changes. So far i have started with adrenaline: making patch for the sceGe module, defining interface between vita and psp, and now i am working on porting GPU statemachine and drawing from ppsspp to vita.

I will probably make it standalone test-app, to test whether it can display anything before integrating it into pspemu.

**Comment by robots** [comment]
BTW: to better understand how sceGe works, I had to improve uofw's sceGe implementation:

https://github.com/uofw/uofw/pull/58

My implementation will use same displaylist management. I have changed accesses to hardware and replaced them with request in similar fashion as Adrenaline does it:

https://github.com/TheOfficialFloW/Adrenaline/blob/master/cef/systemctrl/adrenaline.c#L55

I have not yet settled on decision whether this will be new module, or runtime patch to original sceGe. Runtime patch "could" be turned off, when original Ge is to be used.

**Comment by devnoname120** [comment]
Super excited that uOFW was useful for something :)

**Comment by GamerzHell9137** [comment]
@robots Would there be any possibilities to get a fast forward options for PSP games just like Adrenaline has right now with PSX games?

**Comment by robots** [comment]
I don't really know. I have not researched it. Native resolution is enough work already :-)

**Comment by ghost** [comment]
This will make the XMB at the native resolution too?

I'm a xmb modder in the ps3 scene and wanted to port it to psp xmb, but i only have a ps vita and it while porting the things are pixelated due to the strecht of the psvita screen, so it will be running at native too?

**Comment by Fershock** [comment]
Any advance?

**Comment by robots** [comment]
Right now this project is on hold as some other high priority project in my life is nearing very tough decision - either it works or not and 3 years of R&D are trashed. Have patience please :-)

**Comment by and-war** [comment]
These things take time, would be amazing to head PSP support too, playing PSPs games in hi-res and faster let's just wait and give robots time

**Comment by orestisns** [comment]
Hi @robots 
Any progress ?

**Comment by Fershock** [comment]
I think this is one of the most ambitious and interesting projects on the PSVita scene.  I hope it can be finished

**Comment by Invictaz** [comment]
I think we need more developers than 1 to finish it this year.

**Comment by Nacho-exe** [comment]
not to say i have much experience with creating vita hacks, but didnt TheFloW achieve this with the gta psp titles running at the vitas native resolution? I'd assume you could reverse engineer those hacks to work with other psp titles weather it be required to make a new hack for each game or a universal one most likely for adrenaline. also another idea is to convert psp titles into ps vita format, therefore making the psp (now technically vita) titles essentially bypassing the psp cpu + gpu and using the vita ones? im probably fantasizing about how it could be possible but to me this seems the best option.

**Comment by Mar2ck** [comment]
TheFlow's method was just changing the game's internal resolution and reducing the color so it would fit in the psp's vram. It still uses the psp's GPU so its still limited to what a PSP can render. What robots is doing is moving the rendering from the internal psp to the vita so it can be rendered using the Vita's more powerful hardware so it would work with way more games and would be full color. 

You can't simply convert PSP games to vita format either unfortunately. PSP is MIPS, Vita is ARM. You'd need the source code of PSP games to recompile them as Vita games

**Comment by Fershock** [comment]
> not to say i have much experience with creating vita hacks, but didnt TheFloW achieve this with the gta psp titles running at the vitas native resolution? I'd assume you could reverse engineer those hacks to work with other psp titles weather it be required to make a new hack for each game or a universal one most likely for adrenaline. also another idea is to convert psp titles into ps vita format, therefore making the psp (now technically vita) titles essentially bypassing the psp cpu + gpu and using the vita ones? im probably fantasizing about how it could be possible but to me this seems the best option.

Could you help in the project? I think the more people working on this project the better

**Comment by Fershock** [comment]
> TheFlow's method was just changing the game's internal resolution and reducing the color so it would fit in the psp's vram. It still uses the psp's GPU so its still limited to what a PSP can render. What robots is doing is moving the rendering from the internal psp to the vita so it can be rendered using the Vita's more powerful hardware so it would work with way more games and would be full color

Do you know if @robots could progress or the project is paused right now?

**Comment by Nacho-exe** [comment]
> Could you help in the project? I think the more people working on this project the better

if someone gave me a crash course with everything i needed to know then id be happy to help, i have wayy too much free time on my hand to not put my 2 cents in the ring. But again the issue is i dont know enough about how the vita works to be much of a contributor.

**Comment by robots** [comment]
Crash course is understanding how psp's gpu works. There is almost no documentation.

MY work is paused for now, it needs to get lighter at work, too much work with new POS system as laws changed recently. But i still want to make this work :-) dont worry !

**Comment by Nacho-exe** [comment]
> But i still want to make this work :-) dont worry !

let me know if theres any way myself, or anyone else here can help with anything. Im sure we would all love to help out

**Comment by smokingblaze** [comment]
Best of luck to the project!

**Comment by robots** [comment]
i have restarted work on this thing :-) It will take some time to load all the information back to head, and put them into code, please be patient

**Comment by cfeviper** [comment]
is this also gonna make ps1 games look better when using adrenaline

**Comment by devnoname120** [comment]
> is this also gonna make ps1 games look better when using adrenaline

https://github.com/vita-nuova/bounties/issues/5#issuecomment-525238119

**Comment by cfeviper** [comment]
> > is this also gonna make ps1 games look better when using adrenaline
> 
> [#5 (comment)](https://github.com/vita-nuova/bounties/issues/5#issuecomment-525238119)

is it already  possible then to play psx games at a higher resolution

**Comment by Fershock** [comment]
> i have restarted work on this thing :-) It will take some time to load all the information back to head, and put them into code, please be patient

Good luck!!

**Comment by darkskyVN** [comment]
How long this project will take to compelete?

**Comment by AceCuba** [comment]
did you need any betatester ?

**Comment by Fershock** [comment]
Hey @robots , how's your progress going? :)

**Comment by Invictaz** [comment]
I think Robots won't reply because it takes a lot more work than a month to complete this. We might check back at the end of summer, or maybe the end of the year 2020.

**Comment by tommapar** [comment]
Hey, sorry if what I'm saying is pretty ignorant, but wouldn't it be theoretically easy to maybe port PPSSPP to the Vita? I mean it's already on Android, and works pretty great, 

If I'm not mistaken most Android phones use ARM CPUs as well. Wouldn't it be easier to do it from that angle instead of translating every hardware call from scratch? I mean allegedly, Sony made it easy to make Android Vita ports.

**Comment by robots** [comment]
Hi,
Sorry to reply so late. I am afraid i dont have any free time to work on this project. I thought after christmas the time at work will be little more loose. But its not happening. I can't find a way to allocate time to this project. Again sorry for that.

**Comment by Mar2ck** [comment]
@tommapar The point of this is to port just the graphics portion of PPSSPP to vita because theres no point emulating the psp's cpu when the vita already has one inside

**Comment by Mar2ck** [comment]
@robots 
Thank you for your work! Would it be possible to publish what you've managed to do so far?

**Comment by tommapar** [comment]
@Mar2ck Ok I think I get it. @robots Aww man, that sucks. Still, thanks for your contributions and godspeed in your other endeavors :)

**Comment by Fershock** [comment]
No problem @robots , it is normal to have problems to combine job with this.  What you could ask for is development assistance.  I would propose to create a Reddit post in both vitapiracy and vitahacks to see if anyone is interested in participating in the project.

**Comment by polyjitter** [comment]
@robots completely understand where you are coming from with the job issues, hope things get better for you from here on out. 

It would benefit everyone greatly, however, if you were to potentially post your current work and findings in some format for others to be able to continue and benefit from, as is the benefits of opening source. 

It sounds like you'd gotten quite far with your work and research, and I fear if the community has to start completely from scratch again, this may never be achieved!

**Comment by L0v3daddy** [comment]
Hope someone can keep up the project from where @robots left. The GTA games look much better in native res and I can imagine a buch of games that would look just like Vita games with the upres (ghost of sparta and peace walker come to mind).

**Comment by Invictaz** [comment]
@L0v3daddy With all due respect but the plan for this was already suggested earlier:


1.  patch PSP's sceGe module to redirect Texture memory/DrawList memory to shared Vita memory instead of GE's private memory
2. port PPSSPP's GE implementation to Vita - this step will not include scaling for sake of simplicity
3. add scaling to GE implementation once 2. works correctly.

So we can work on step one, if we have the knowledge from Vita3K. I suggest we set up a separate Github for this.

**Comment by bezza191** [comment]
@Invictaz If a new git is set up for going in this direction, could you please drop details in here (assuming @robots and @Epsilon1033 are ok with this) as I am looking to follow progress, as im sure others are also

**Comment by L0v3daddy** [comment]
@Invictaz right, that was clarified a while ago, I only mentioned Robots because he seemed to be the only one that was actively working on the project (at least from the perspective of a bystander).

**Comment by Invictaz** [comment]
I won't set it up since I'm not a coder myself

**Comment by L0v3daddy** [comment]
I wonder if this bounty can get high enough to interest someone like The Flow...

**Comment by AlJoelson** [comment]
> I wonder if this bounty can get high enough to interest someone like The Flow...

It is a heck of a lot of money at this point!

**Comment by evertonstz** [comment]
> 
> 
> > I wonder if this bounty can get high enough to interest someone like The Flow...
> 
> It is a heck of a lot of money at this point!

With all due respect to both and everyone that donated, if someone ends up closing this bounty it'll be done due to love for the system, the money is symbolic since it's a lot of high skilled work, research and specially time involved.

**Comment by IridescentRose** [comment]
I'd be interested, but I don't have a Vita, nor the time with multiple projects. However, I am a PSP developer and kinda interested. Based on the data we have, couldn't we just stream GPU opcodes directly to a buffer in Vita memory and then implement an API based on that? That way you basically intercept the output to the GPU and just implement that (which you could do in PPSSPP or just roll-your-own) since the GPU opcodes are better documented: http://hitmen.c02.at/files/yapspd/psp_doc/chap11.html#sec11. If one did just write an interpreter on the Vita, then you wouldn't need the step 3 of scaling, since you'll use GXM anyways and draw and scale things similarly.

Tl;Dr:

1.) Divert GPU opcodes to vita memory (take the outgoing bus and push it to adrenaline) instead of dumping GE RAM
2.) Interpret GPU opcodes to GXM commands

**Comment by L0v3daddy** [comment]
Looking forward to see what other developers have to say to that possibility, since I'm just an enthusiast. Just out og curiosity, don't you know any other psp developer that may be interested?

**Comment by IridescentRose** [comment]
I'm part of the PSP Homebrew Discord Server. Nobody else really was interested. I frankly am not a Vita dev nor intend to be one. But I do want to learn more about the GE and GPU system.

**Comment by robots** [comment]
So, since this corona stuff, work is being relaxed. I think i will have time to work on this.

I have created repository, and pushed all my stuff into that. I cannot push my notes there as they are in my head :-)

https://github.com/robots/Adrenaline/commit/a0b1b68b25f1f7f8e783fd05d7e533e2a815433e

Description:
psp part: this is copied and reworked uofw ge driver. All writes and reads from registers have been replaced by requests to vita.

https://github.com/robots/Adrenaline/blob/master/cef/systemctrl/ge-vita.c#L78

PSP ge driver is source:
https://github.com/uofw/uofw/tree/master/src/ge
with my additions (hopefully fixes):
https://github.com/uofw/uofw/pull/58/files


Since there is limited number of shared resources between vita and psp, we have to share. There is at least one messagebox free (KERMIT_MODE_EXTRA_1, adrenaline uses KERMIT_MODE_EXTRA_2). But these are no interrupts free. We will share the interrupt adrenaline uses. Adrenaline uses this only to signal save/load of state and memory card reinsert - these are very rare interrupts, thus Adrenaline will not mind sharing.

@NT-Bourgeois-Iridescence-Technologies  it is not possible just to stream gpu data from psp. Gpu data is in ram that is not accessible from vita.

I patch the sceGeEdramGetAddr function to return address in vita's CDRAM (mapped into PSP's memory space) This way PSP game will write directly to shared memory. This makes it harder to have both gpu on vita, and ge on psp work at the same time (i would like to have switch, for games that are not yet compatible, but i am sure we can unpatch the ge driver or add some flag to select code.)
In psp code i think it misses just the hooking stuff (and the flag)

For the vita part, i have started to write the GE state machine in C. I use PPSSPP's gpu framework as source of ideas. But its like 2percent of work done.

The roadmap should be something like:
- write vita's ge statemachine
- test it on vita only (with some psp example from https://github.com/pspdev/pspsdk/tree/master/src/samples/gu)
- finish psp ge driver hooking.
- test complete system.

So what is needed is one PSP developer who knows psp hooking stuff to finish the ge driver.

**Comment by polyjitter** [comment]
@robots Thank you for your amazing work! This is really promising to see. Hope someone else sees this and is interested :)

**Comment by eleqian** [comment]
@robots I'm interested in this project. I‘ve tried ge driver hooking some days ago, two ways are ok:
1. HIJACK_FUNCTION(FindProc("sceGE_Manager", "sceGe_driver", 0XAB49E76A), sceGeListEnQueue_patched, sceGeListEnQueue_org);
2. sctrlHENPatchSyscall(K_EXTRACT_IMPORT(&sceGeListEnQueue), sceGeListEnQueue_patched);

**Comment by L0v3daddy** [comment]
Super happy to see the project being picked up again, thank you for your work and love for the system!

**Comment by L0v3daddy** [comment]
What happened to the bountysource page? I can't access it.

**Comment by and-war** [comment]
Can ppl just stop spamming here asking the devs, yes this is cool, but they need time, and by asking will not speed things up, lets all just wait and soon we will have amazing features on the good old Vita.

**Comment by XDeadInside** [comment]
Hey everyone, I've been following this issue since last months... I don't wanna disturb anyone, just want to thank the developers working on this, especially @robots, I love vita and I really appreciate your work. ❤🙇‍♂️

**Comment by bielgiroto** [comment]
@XDeadInside I was going to say the same thing XD
I've been praying every night for this.

**Comment by evertonstz** [comment]
For the love of God, people, if you you're not coding or contributing, don't spam here, Github isn't Reddit. If you wanna show your appreciation or how much you want this feature, go and donate to the bountie.
This should be as tight as possible so people that are working or looking to work on the issue don't need to scroll thought dozens of unrelated posts to get some information.

I would suggest Rinnegatamante to give this thread a good cleanup, including this post obviously.

**Comment by XDeadInside** [comment]
I would do if I could, it's not really that easy to donate anything thanks to US restrictions over my country, we don't access to credit cards and if we do, it usually gets totally banned... I just wanted to show my appreciation at least... sorry

**Comment by thegarlicman123** [comment]
$2000

**Comment by StamKaly** [comment]
[GePatch](https://github.com/TheOfficialFloW/GePatch) is also in production which seems to have a few games already working.

**Comment by BreakPoints** [comment]
> [GePatch](https://github.com/TheOfficialFloW/GePatch) is also in production which seems to have a few games already working.

GePatch halted production when TheFloW moved to GTA SA on Vita

**Comment by Windows-10-Is-Shit** [comment]
> The feasibility has already been discussed on reddit so it is certainly possible

Well, if it's been discussed on reddit and it's certainly possible then why isn't reddit doing this? 😄

**Comment by wefalltomorrow** [comment]
Has any progress been made with this?

**Comment by Girotin** [comment]
> Has any progress been made with this?

Since TheFlow gave up GEpatch, I believe nobody else picked up the project.

I won't judge it though; even if we have quite a good money already accumulated, idk if it's enough for a dev/group of devs to see the amount of job needed and think: "hmm, I think I'll try doing this", this is a VERY ambicious project.

Maybe if we reach 5000 someday? 10000? Idk how much would be necessary for a dev/devs to get interested in this XD
But seeing thinks how they are, 2000 still isn't enough :v

( If there's someone working on this, feel free to correct me )
